backends/p4tools/testgen/lib/execution_state.h

```diff
@@ -0,0 +1,476 @@
+
+    /// The program trace for the current program point (i.e., how we got to the current state).
+    std::vector<gsl::not_null<const TraceEvent*>> trace;
+
+    /// List of visited statements. Used for code coverage.
+    std::vector<const IR::Statement*> visitedStatements;
+
+    /// The remaining body of the current function being executed.
+    ///
+    /// Invariant: if this is empty, then so is the @stack, and this state is terminal.
+    Continuation::Body body;
+
+    /// When execution of the @body completes, the @stack is popped and execution passes to the
+    /// normal continuation in the popped frame. Any returned value is given as the argument to
+    /// that continuation. If the current @body throws an exception, the stack is popped until we
+    /// find a corresponding handler. The handler body becomes the new @body, and the frame below
+    /// becomes the top of the stack.
+    ///
+    // Invariant: if the @body is empty, then so is this, and this state is terminal.
+    std::stack<gsl::not_null<const StackFrame*>> stack;
+
+    /// State properties are bools, integers, or strings that can be set and propagated across
+    /// execution state. They are used to influence execution along a particular continuation path.
+    /// For example, we can set the "inUndefinedState" property to indicate that all state being
+    /// written while this variable is active is tainted. This property must be unset manually to
+    /// resume normal operation. Usually, the unset call is inserted directly after the tainted
+    /// sequence of commands has executed.
+    std::map<cstring, Continuation::StateProperty> stateProperties;
+
+    // Test objects are classes of variables that influence the execution of test frameworks. They
+    // are collected during interpreter execution and consumed by the respective test framework. For
+    // example, when visiting a table, the interpreter will generate a table test object per table,
+    // which defines control plane match action entries. Once the interpreter has solved for the
+    // variables used by these test objects and concretized the values, they can be used to generate
+    // a test. Test objects are not constant because they may be manipulated by a target back end.
+    std::map<cstring, std::map<cstring, const TestObject*>> testObjects;
+
+    /// The parserErrorLabel is set by the parser to indicate the variable corresponding to the
+    /// parser error that is set by various built-in functions such as verify or extract.
+    const IR::Member* parserErrorLabel = nullptr;
+
+    /// This variable tracks how much of the input packet has been parsed.
+    /// Typically, this is equivalent to the size of the input packet. However, there may be
+    /// functions such as lookahead that peek into the program packet without actually advancing the
+    /// cursor. The cursor ensures that parsing remains consistent in these cases.
+    int inputPacketCursor = 0;
```

Comment:
why not size_t?

Reply:
Because this variable is also incremented by `width_bits()`, which has the return type `int`. It might make sense to return `size_t` for `width_bits` in general. This is a very common type mismatch in the compiler.